apis/projectcontour/v1/detailedconditions.go

```diff
@@ -0,0 +1,85 @@
+	// +kubebuilder:validation:MaxLength=316
+	Type string `json:"type" protobuf:"bytes,1,opt,name=type"`
+	// status of the condition, one of True, False, Unknown.
+	// +required
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Enum=True;False;Unknown
+	Status ConditionStatus `json:"status" protobuf:"bytes,2,opt,name=status"`
+	// reason contains a programmatic identifier indicating the reason for the condition's last transition.
+	// Producers of specific condition types may define expected values and meanings for this field,
+	// and whether the values are considered a guaranteed API.
+	// The value should be a CamelCase string.
+	// This field may not be empty.
+	// +required
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MaxLength=1024
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:Pattern=`^[A-Za-z]([A-Za-z0-9_,:]*[A-Za-z0-9_])?$`
+	Reason string `json:"reason" protobuf:"bytes,3,opt,name=reason"`
+	// message is a human readable message indicating details about the transition.
+	// This may be an empty string.
+	// +required
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MaxLength=32768
+	Message string `json:"message" protobuf:"bytes,4,opt,name=message"`
+}
+
+// DetailedCondition is an extension of the normal Kubernetes conditions, with two extra
+// fields to hold sub-conditions, which provide more detailed reasons for the state (True or False)
+// of the condition.
+// `errors` holds information about sub-conditions which are fatal to that condition and render its state False.
+// `warnings` holds information about sub-conditions which are not fatal to that condition and do not force the state to be False.
+// Remember that Conditions have a type, a status, and a reason.
+// The type is the type of the condition, the most important one in this CRD set is `Valid`.
+// In the case of `Valid`, `status: true` means that the object is has been ingested into Contour with no errors.
+// `warnings` may still be present, and will be indicated in the Reason field.
+// `Valid`, `status: false` means that the object has had one or more fatal errors during processing into Contour.
+//  The details of the errors will be present under the `errors` field.
+// There should never be subconditions under `errors` when `status` is `true`.
+type DetailedCondition struct {
+	// TODO(youngnick): Replace this with metav1.Condition once we have moved to a client-go
+	// version that includes it. Also includes deleting kubeconditions.go.
+	Condition `json:",inline"`
+	// Errors contains a slice of relevant error subconditions for this object.
+	// Subconditions are expected to appear when relevant (when there is a error), and disappear when not relevant.
+	// An empty slice here indicates no errors.
+	Errors []SubCondition `json:"errors,omitempty"`
```

Comment:
Use `+optional` for both `Errors` and `Warnings` fields?

Reply:
Yes, I intended to, thanks.